app/Http/Controllers/ExpenseController.php

```diff
@@ -0,0 +1,146 @@
+{
+    /**
+     * Display a listing of the resource.
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function index()
+    {
+        $expenses = $this->getExpenses();
+        return view('welcome', compact('expenses'));
+    }
+
+    /**
+     * Show the form for creating a new resource.
+     *
+     * @return \Illuminate\Http\Response
+     */
+    public function create()
+    {
+        //
+    }
+
+    /**
+     * Store a newly created resource in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @return \Illuminate\Http\Response
+     */
+    public function store(Request $request)
+    {
+        $validate = $request->validate([
+            'name' => 'required|min:3',
+            'entry_date' => 'required|date_format:Y-m-d H:i:s',
+            'amount' => 'required|numeric',
+        ]);
+
+        $expense = new Expense();
+        $expense->name = $validate['name'];
+        $expense->entry_date = $validate['entry_date'];
+        $expense->amount = $validate['amount'];
+        $expense->save();
+
+        return response()->json($this->getExpenses());
+    }
+
+    /**
+     * Display the specified resource.
+     *
+     * @param  \App\Expense  $expense
+     * @return \Illuminate\Http\Response
+     */
+    public function show(Expense $expense)
+    {
+        //
+    }
+
+    /**
+     * Show the form for editing the specified resource.
+     *
+     * @param  \App\Expense  $expense
+     * @return \Illuminate\Http\Response
+     */
+    public function edit(Expense $expense)
+    {
+        //
+    }
+
+    /**
+     * Update the specified resource in storage.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \App\Expense  $expense
+     * @return \Illuminate\Http\Response
+     */
+    public function update(Request $request, Expense $expense)
+    {
+        $expense->update([
+            'name' => $request->input('name'),
+            'entry_date' => $request->input('entry_date'),
+            'amount' => $request->input('amount'),
+        ]);
+
+        return response()->json($expense);
+    }
+
+    /**
+     * Remove the specified resource from storage.
+     *
+     * @param \App\Expense $expense
+     * @return \Illuminate\Http\Response
+     * @throws \Exception
+     */
+    public function destroy(Expense $expense)
+    {
+        $expense->delete();
+    }
+
+    public function getCsvRowCount(Request $request)
+    {
+        $validate = $request->validate([
+            'file' => 'required|file|mimes:csv,txt',
+        ]);
+
+        $count = (new ExpenseImport)->toCollection($validate['file'])->shift()->count();
+
+        return response()->json([
+            'rows' => $count,
+        ]);
+    }
+
+    public function import(Request $request)
+    {
+        $validate = $request->validate([
+            'file' => 'required|file|mimes:csv,txt',
+        ]);
+
+        Excel::import(new ExpenseImport, $validate['file']);
```

Comment:
This should probably be run in a queue job to accommodate for processing massive files.